### PR TITLE
solution

### PR DESCRIPTION
**Pull request by maksym-kostetskyi**

### CODE REVIEW

File: src/transformStateWithClones.js [hunk 1]
```diff
@@ -6,8 +6,34 @@
  *
  * @return {Object[]}
  */
+
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateVers = [state];
+
+  for (let i = 0; i < actions.length; i++) {
```
Comment (ivbutsykin) [inline src/transformStateWithClones.js:13]: You can use `for of` instead `for` here

File: src/transformStateWithClones.js [hunk 2]
```diff
@@ -6,8 +6,34 @@
  *
  * @return {Object[]}
  */
+
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateVers = [state];
```
Comment (ivbutsykin) [inline src/transformStateWithClones.js:11]: So, as I understand, at first you add `state` in array and after that you remove it from array when make return `stateVers.slice(1)`.

It's seems not good. I can suggest you make sth like that:
```suggestion
  const result = [];
  const stateCopy = { ...state };
```

When you will push your data in result and have not modified state.

File: src/transformStateWithClones.js [hunk 3]
```diff
@@ -6,8 +6,34 @@
...
+
+  for (let i = 0; i < actions.length; i++) {
+    let newState = { ...stateVers[i] };
+
+    switch (actions[i].type) {
+      case 'addProperties':
+        for (const [key1, value1] of Object.entries(actions[i].extraData)) {
+          newState[key1] = value1;
+        }
```
Comment (ivbutsykin) [inline src/transformStateWithClones.js:20]: It can be simplified: use `Object.assign`

File: src/transformStateWithClones.js [hunk 4]
```diff
@@ -6,8 +6,34 @@
...
+        break;
+      case 'removeProperties':
+        for (const key2 of actions[i].keysToRemove) {
+          if (newState.hasOwnProperty(key2)) {
```
Comment (ivbutsykin) [inline src/transformStateWithClones.js:24]: I think, if you will delete undefined value it's will still work, so you can remove this check

**Review by ivbutsykin — CHANGES_REQUESTED** [review]

**Review by slavanosatskyi — APPROVED** [review]
all good, but don't forget about default case in switch case statement

**Pull request closed without merging**